core/modules/replica/WorkerServerConnection.h

```diff
@@ -0,0 +1,225 @@
+///
+/// class WorkerServerConnection
+/// (see individual class documentation for more information)
+
+// System headers
+#include <memory>
+
+// Third party headers
+#include <boost/asio.hpp>
+
+// Qserv headers
+#include "proto/replication.pb.h"
+#include "replica/ProtocolBuffer.h"
+#include "replica/ServiceProvider.h"
+
+// This header declarations
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+// Forward declarations
+class WorkerProcessor;
+
+/**
+  * Class WorkerServerConnection is used for handling connections from
+  * remote clients. One instance of the class serves one client at a time.
+  *
+  * Objects of this class are inistantiated by WorkerServer. After that
+  * the server calls this class's method beginProtocol() which startes
+  * a series of asynchronous operations to communicate with remote client.
+  * When all details of an incoming request are obtained from the client
+  * the connection object forwards this request for actual processing
+  * to an instace of the WorkerProcessor class. A response reseived from
+  * the processor is serialized and sent back (asynchroniously) to
+  * the client.
+  */
+class WorkerServerConnection
+    :   public std::enable_shared_from_this<WorkerServerConnection> {
+
+public:
+
+    /// Shared pointer type for the class
+    typedef std::shared_ptr<WorkerServerConnection> Ptr;
+
+    /**
+     * Static factory method is needed to prevent issue with the lifespan
+     * and memory management of instances created otherwise (as values or via
+     * low-level pointers).
+     *
+     * @param serviceProvider - provider of various services
+     * @param processor       - processor of long requests
+     * @param io_service      - enpoint for network I/O
+     *
+     * @return pointer to the new object created by the factory
+     */
+    static Ptr create(ServiceProvider::Ptr const& serviceProvider,
+                      WorkerProcessor& processor,
+                      boost::asio::io_service& io_service);
+
+    // Default construction and copy semantics are prohibited
+
+    WorkerServerConnection() = delete;
+    WorkerServerConnection(WorkerServerConnection const&) = delete;
+    WorkerServerConnection& operator=(WorkerServerConnection const&) = delete;
+
+    ~WorkerServerConnection() = default;
+
+    /// @eturn network socket associated with the connection
+    boost::asio::ip::tcp::socket& socket() { return _socket; }
+
+    /**
+     * Begin communicating asynchroniously with a client. This is essentially
+     * an RPC protocol which runs in a loop this sequence of steps:
+     * 
+     *   - ASYNC: read a frame header of a request
+     *   -  SYNC: read the request header (request type, etc.)
+     *   -  SYNC: read the request body (depends on a type of the request) 
+     *   - ASYNC: write a frame header of a reply to the request
+     *            then write the reply itself
+     *
+     * NOTES: A reason why the read phase is split into three steps is
```

Comment:
actually four steps?

Reply:
Fixe. The initial implementation had 3 steps because of no separation between the header and the body in the protocol. The split happened latter.
BTW: that's an illustration why @fritzm don't like Doxygen :-)